### PR TITLE
Add match and missing warnings in Test results

### PR DESCRIPTION
**Pull request by Cheshiriks**
What's done:
* Added match and missing warnings in Test results

### CODE REVIEW

File: save-common/src/commonMain/kotlin/org/cqfn/save/core/result/TestResult.kt [hunk 1]
```diff
@@ -14,10 +14,14 @@ import kotlinx.serialization.UseSerializers
...
  * @property debugInfo additional info that can be set during execution. Might be absent.
+ * @property missing number of missing warnings
+ * @property match number of match warnings
  */
 @Serializable
 data class TestResult(
     val resources: Plugin.TestFiles,
     val status: TestStatus,
     val debugInfo: DebugInfo? = null,
+    val missing: Int? = null,
```
Comment (petertrr) [inline save-common/src/commonMain/kotlin/org/cqfn/save/core/result/TestResult.kt:25]: I feel like such information pollutes this class a little. It's warn-plugin-specific, while `TestResult` is shared class for all plugins. It's too low-level for `TestResult`, imo. I'd rather see it in `debugInfo` or maybe even in a special subclass of `DebugInfo` specifically for warn-plugin

Comment (Cheshiriks) [reply save-common/src/commonMain/kotlin/org/cqfn/save/core/result/TestResult.kt:25]: Yes. Agree)

Comment (Cheshiriks) [reply save-common/src/commonMain/kotlin/org/cqfn/save/core/result/TestResult.kt]: Fixed

File: save-common/src/commonMain/kotlin/org/cqfn/save/core/result/DebugInfo.kt [hunk 2]
```diff
@@ -10,11 +12,23 @@ import kotlinx.serialization.Serializable
...
+)
+
+/**
+ * @property missing number of missing warnings
+ * @property match number of match warnings
+ */
+@Serializable
+data class CountWarnings(
+    val missing: Int? = null,
+    val match: Int? = null,
```
Comment (petertrr) [inline save-common/src/commonMain/kotlin/org/cqfn/save/core/result/DebugInfo.kt:33]: I think here these fields can be non-nullable. If this info is not available, `CountWarnings` instance won't be created at all.

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt [hunk 3]
```diff
@@ -26,7 +26,7 @@ class ResultsChecker(
      * @return [TestStatus]
      */
     @Suppress("TYPE_ALIAS")
-    internal fun checkResults(testFileName: String): TestStatus {
+    internal fun checkResults(testFileName: String): Triple<TestStatus, Int, Int> {
```
Comment (petertrr) [inline save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt:29]: I think `Pair<TestStatus, CountWarnings>` would be more readable

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt [hunk 4]
```diff
@@ -39,25 +39,28 @@ class ResultsChecker(
...
         val unexpectedWarnings = actualWarnings - actualMatchedWithExpectedWarnings
 
+        val missing = missingWarnings.size
+        val match = expectedWarningsMatchedWithActual.size
+
         return when (missingWarnings.isEmpty() to unexpectedWarnings.isEmpty()) {
-            false to true -> Fail(
+            false to true -> Triple(Fail(
                 "$MISSING $missingWarnings",
                 "$MISSING (${missingWarnings.size}). $MATCHED (${expectedWarningsMatchedWithActual.size})"
```
Comment (petertrr) [inline save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt:48]: `missing` and `match` coutners can be used everywhere here too

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt [hunk 5]
```diff
@@ -39,25 +39,28 @@ class ResultsChecker(
         val missingWarnings = expectedWarnings - expectedWarningsMatchedWithActual
         val unexpectedWarnings = actualWarnings - actualMatchedWithExpectedWarnings
 
+        val missing = missingWarnings.size
+        val match = expectedWarningsMatchedWithActual.size
```
Comment (petertrr) [inline save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt:43]:
```suggestion
        val matched = expectedWarningsMatchedWithActual.size
```

File: save-common/src/commonMain/kotlin/org/cqfn/save/core/result/DebugInfo.kt (hunk 2 above)
Comment (Cheshiriks) [reply save-common/src/commonMain/kotlin/org/cqfn/save/core/result/DebugInfo.kt]: Fixed

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt (hunk 3 above)
Comment (Cheshiriks) [reply save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt]: Fixed

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt (hunk 4 above)
Comment (Cheshiriks) [reply save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt]: Fixed

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt (hunk 5 above)
Comment (Cheshiriks) [reply save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt]: Fixed

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt [hunk 6]
```diff
@@ -39,25 +40,28 @@ class ResultsChecker(
...
+
         return when (missingWarnings.isEmpty() to unexpectedWarnings.isEmpty()) {
             false to true -> Fail(
                 "$MISSING $missingWarnings",
-                "$MISSING (${missingWarnings.size}). $MATCHED (${expectedWarningsMatchedWithActual.size})"
-            )
-            false to false -> createFailFromDoubleMiss(missingWarnings, unexpectedWarnings, expectedWarningsMatchedWithActual)
-            true to true -> Pass("$ALL_EXPECTED (${expectedWarningsMatchedWithActual.size})")
+                "$MISSING ($missing). $MATCHED ($matched)"
+            ) to CountWarnings(missing, matched)
```
Comment (petertrr) [inline save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt:50]: `to CountWarnings(missing, matched)` is duplicated in every branch, please move it into a single place

**Review by petertrr — CHANGES_REQUESTED** [review]

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt (hunk 6 above)
Comment (Cheshiriks) [reply save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt]: Fixed

**Review by petertrr — APPROVED** [review]

**Comment by petertrr** [comment]
Lol, we've hit API usage limit here too:
```
Egress is over the account limit. (HTTP status code: 503, URL: https://github.com/cqfn/diKTat/releases/download/v1.0.0-rc.2/diktat-1.0.0-rc.2.jar
```
I've created https://github.com/diktat-static-analysis/save-cloud/issues/462 in save-cloud for the same reason, we should add this in save too.

**Pull request merged**